### ISSUE TITLE
Multiple instances of custom metrics don't behave as expected with multiple output model when evaluating.

### ISSUE THREAD

**Issue by dithyrambe** [tensorflow/tensorflow#43699]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): **YES**
- OS Platform and Distribution: **Ubuntu 20.04.1 LTS**
- TensorFlow installed from (source or binary): `pip install`
- TensorFlow version (use command below): `tensorflow==2.3.0`
- Python version: **Python 3.6.9 :: Anaconda, Inc.**

**Current behavior**
Custom metrics behave unexpectedly when applied on a model / dataset with multpiple outputs.
Below is a snippet to reproduce.

**Describe the expected behavior**
Results should be consistent whatever the number of outputs.

**Standalone code to reproduce the issue**
```python
import numpy as np
import tensorflow as tf


def build_dataset(n_outputs):
    """
    Build a random dataset.
    Target is a dictionnary of `n_output` elements with keys as ["out0", "out1", ...]
    """
    input_data = np.random.normal(size=(1024, 5)).astype("float32")
    target_data = {
        f"out{n}": np.random.normal(size=(1024,)).astype("float32")
        for n in range(n_outputs)
    }
    input_dataset = tf.data.Dataset.from_tensor_slices(input_data)
    target_dataset = tf.data.Dataset.from_tensor_slices(target_data)
    dataset = tf.data.Dataset.zip((input_dataset, target_dataset)).batch(16)
    return dataset


class Model(tf.keras.Model):
    """Model with `n_outputs` outputs"""
    def __init__(self, n_outputs):
        tf.keras.Model.__init__(self)
        self.hidden = tf.keras.layers.Dense(16)
        self.out = {f"out{n}": tf.keras.layers.Dense(1) for n in range(n_outputs)}

    def call(self, inputs):
        hidden = self.hidden(inputs)
        return {out_name: out_layer(hidden) for out_name, out_layer in self.out.items()}


class DumbMetric(tf.keras.metrics.Metric):
    """This metrics takes a param at __init__ time and then always return param as result"""
    def __init__(self, name="Dumb", param=0., **kwargs):
        tf.keras.metrics.Metric.__init__(self, name=name, **kwargs)
        self.param = float(param)

    def update_state(self, y_true, y_pred, sample_weight):
        pass

    def result(self):
        return self.param


for n in (1, 2):
    # For 1 output, metrics behave as expected:
    #   - metric1 returns 1
    #   - metric2 returns 2
    #
    # For 2 outputs, metrics don't behave as expected:
    #   - metric1 returns 0 (default `param` value)
    #   - metric2 returns 0 (default `param` value)

    model = Model(n)
    model.compile(
        metrics=[
            DumbMetric(name="metric1", param=1.),
            DumbMetric(name="metric2", param=2.)
        ]
    )
    metrics = model.evaluate(build_dataset(n), return_dict=True, verbose=0)
    print(f"For {n} output(s): ", metrics)
```

**Actual output:**
```
For 1 output(s):  {'loss': 0.0, 'metric1': 1.0, 'metric2': 2.0}
For 2 output(s):  {'loss': 0.0, 'out0_metric1': 0.0, 'out0_metric2': 0.0, 'out1_metric1': 0.0, 'out1_metric2': 0.0}
```

**Excpected output:**
```
For 1 output(s):  {'loss': 0.0, 'metric1': 1.0, 'metric2': 2.0}
For 2 output(s):  {'loss': 0.0, 'out0_metric1': 1.0, 'out0_metric2': 2.0, 'out1_metric1': 1.0, 'out1_metric2': 2.0}
```

**Comment by dithyrambe** [comment]
Hi again,

After further investigation, overriding the following method in `DumbMetric` fixes the issue:

```python
    def get_config(self):
        return {
            **tf.keras.metrics.Metric.get_config(self),
            "param": self.param
        }
```

This is due because metric objects are being copied [through this method](https://github.com/tensorflow/tensorflow/blob/7ea86e9de8a5b6426e7291e0e5477ddaee83ba88/tensorflow/python/keras/engine/compile_utils.py#L487-L489) when multiple outputs are [out there](https://github.com/tensorflow/tensorflow/blob/7ea86e9de8a5b6426e7291e0e5477ddaee83ba88/tensorflow/python/keras/engine/compile_utils.py#L92-L94).

It wasn't clear to me this should be done before subclassing `tf.keras.metrics.Metric`. I may have missed some hint in the documentation. Though, wouldn't it be nice to pass `__init__` `kwargs` by default to `tf.keras.metrics.Metric.get_config` in case user defines custom parameters that should be passed again when "copiying" metrics ?

Thanks in advance.

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with TF v2.3 and TF-nightly. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/ed1d48fcc45a609c9bd0512ab39d31b3/43699.ipynb). Thanks!

**Comment by dithyrambe** [comment]
Having `tf.keras.metrics.Metric.get_config` as an `abc.abstractmethod` with an explicit mention in the docstring/documentation is the simplest "fix" I can think of.

**Comment by sachinprasadhs** [comment]
Was able to reproduce your issue in Tf Nightly 2.6.0-dev20210528, please find the gist [here](https://colab.research.google.com/gist/sachinprasadhs/aa03a1cdcaa690e7ef609624433dba81/43699.ipynb). Thanks!

**Comment by SuryanarayanaY** [comment]
Hi @dithyrambe ,

I have observed that you are passing Metrics as a list.This works for single output model.For multi outputs you have to pass the Metrics in `dictionary` form or else `nestedlist`.Please refer to attached documentation [link](https://www.tensorflow.org/api_docs/python/tf/keras/Model#compile) for same. Could you please change the necessary formats and let us know if still having problems.

Thankyou!
